implementations/python/python/ockam/models/model.py

```diff
@@ -290,33 +302,147 @@ def support_tools(self):
     def support_forced_assistant_answer(self):
         return "bedrock" not in self.name and "litellm_proxy" not in self.name
 
-    async def complete_chat(self, messages: List[dict] | List[ConversationMessage], stream: bool = False, **kwargs):
+    async def complete_chat(
+        self,
+        messages: List[dict] | List[ConversationMessage],
+        stream: bool = False,
+        is_thinking: bool = False,
+        **kwargs,
+    ):
+        """
+        Send a chat completion request to the model.
+
+        :param messages: List of messages to send to the model. Each message should be a dictionary with 'role' and 'content' keys.
+        :param stream: When True, the response will be streamed back as it is generated. If False, the full response will be returned at once.
+        :param is_thinking: Must be true when the thinking has already started but not completed yet.
+        :param kwargs: Additional parameters to pass to the model, such as temperature, max_tokens, etc.
+        :rtype The response from the model, either as a full response or a stream of responses.
+        """
         self.logger.info(f"Processing {len(messages)} messages with model '{self.original_name}'")
         self.logger.debug(f"Sending the following messages to the model: {messages} (stream={stream})")
 
-        messages, kwargs = self.prepare_llm_call(messages, **kwargs)
+        messages, kwargs = self.prepare_llm_call(messages, is_thinking, **kwargs)
 
-        response = await self.router().acompletion(self.name, messages=messages, stream=stream, **kwargs)
-        self.logger.debug(f"Got a response from model '{self.original_name}': {response}")
+        if stream:
+            return self._complete_chat_stream(messages, is_thinking, **kwargs)
+        else:
+            return await self._complete_chat(messages, **kwargs)
+
+    async def _complete_chat_stream(self, messages: List[dict], is_thinking: bool, **kwargs):
+        chunks = None
+
+        if _cache_inference:
```

Comment:
Is there an opportunity to refactor with L.380 and maybe L.348 + L.392 / L.410 + L.420?

Reply:
I did deduplicate a little bit, but couldn't do much